File: app/src/main/java/com/example/weatheradvisor/viewmodels/WeatherViewModel.java
package com.example.weatheradvisor.viewmodels;

import android.content.Context;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.weatheradvisor.BuildConfig;
import com.example.weatheradvisor.models.WeatherResponse;
import com.example.weatheradvisor.services.WeatherApiClient;
import com.example.weatheradvisor.helpers.LocationHelper;
import com.example.weatheradvisor.services.WeatherService;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class WeatherViewModel extends ViewModel {
    private final MutableLiveData<WeatherResponse> weatherData = new MutableLiveData<>();
    private final MutableLiveData<String> errorMessage = new MutableLiveData<>();
    private final WeatherService weatherService;

    public WeatherViewModel() {
        weatherService = WeatherApiClient.getInstance();
    }

    public LiveData<WeatherResponse> getWeatherData() {
        return weatherData;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public void fetchWeatherData(Context context) {
        LocationHelper locationHelper = new LocationHelper(context);
        locationHelper.getLastLocation(location -> {
            if (location != null) {
                double lat = location.getLatitude();
                double lon = location.getLongitude();

                weatherService.getWeather(lat, lon, BuildConfig.weatherApiKey, "metric").enqueue(new Callback<WeatherResponse>() {
                    @Override
                    public void onResponse(Call<WeatherResponse> call, Response<WeatherResponse> response) {
                        if (response.isSuccessful() && response.body() != null) {
                            weatherData.postValue(response.body());
                        } else {
                            errorMessage.postValue("Error: " + response.message());
                        }
                    }

                    @Override
                    public void onFailure(Call<WeatherResponse> call, Throwable t) {
                        errorMessage.postValue("API error: " + t.getMessage());
                    }
                });
            } else {
                errorMessage.postValue("Unable to get location.");
            }
        });
    }
}

File: app/src/main/java/com/example/weatheradvisor/ui/WeatherViewModel.kt
//package com.example.weatheradvisor.ui
//
//import androidx.compose.runtime.mutableStateOf
//import androidx.lifecycle.ViewModel
//import androidx.lifecycle.viewModelScope
//import com.example.weatheradvisor.WeatherApiClient
//import com.example.weatheradvisor.helpers.LocationHelper
//import com.example.weatheradvisor.models.WeatherResponse
//import com.example.weatheradvisor.repositories.WeatherDatabase
//import com.example.weatheradvisor.repositories.WeatherData
//import kotlinx.coroutines.launch
//
//class WeatherViewModel(
//    private val locationHelper: LocationHelper,
//    private val weatherDatabase: WeatherDatabase
//) : ViewModel() {
//
//    val weatherData = mutableStateOf<WeatherResponse?>(null)
//
//    init {
//        fetchWeatherData()
//    }
//
//    private fun fetchWeatherData() {
//        locationHelper.getLastLocation { location ->
//            if (location != null) {
//                val lat = location.latitude
//                val lon = location.longitude
//
//                viewModelScope.launch {
//                    try {
//                        val response = WeatherApiClient.getInstance().getWeather(lat, lon, "your_api_key", "metric")
//                        if (response.isSuccessful) {
//                            weatherData.value = response.body()
//                            cacheWeatherData(response.body())
//                        } else {
//                            loadCachedWeatherData()
//                        }
//                    } catch (e: Exception) {
//                        loadCachedWeatherData()
//                    }
//                }
//            }
//        }
//    }
//
//    private fun cacheWeatherData(weatherResponse: WeatherResponse?) {
//        weatherResponse?.let {
//            val weatherData = WeatherData(
//                temperature = it.main.temp,
//                description = it.weather[0].description,
//                timestamp = System.currentTimeMillis()
//            )
//
//            viewModelScope.launch {
//                weatherDatabase.weatherDao().insert(weatherData)
//            }
//        }
//    }
//
//    private fun loadCachedWeatherData() {
//        viewModelScope.launch {
//            val cachedData = weatherDatabase.weatherDao().getLatestWeatherData()
//            cachedData?.let {
//                weatherData.value = WeatherResponse(
//                    main = WeatherResponse.Main(temp = it.temperature),
//                    weather = listOf(WeatherResponse.Weather(description = it.description)),
//                    // попълни и останалите полета, ако са нужни
//                )
//            }
//        }
//    }
//}